### PR TITLE
HPCC-27655 Add support for TimeStart to roxie

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
@richardkchapman bulk of the change is mechanical, adding a start() that calls doStart().  Splitter and library call have different code to record the amount of time for the first start from an output.

**Comment by ghalliday** [comment]
@mckellyln this is the PR to add the TimeStart stat.

File: roxie/ccd/ccdserver.cpp [hunk 1]
```diff
@@ -9138,10 +9153,12 @@ class CRoxieServerThroughSpillActivity : public CRoxieServerActivity
             }
             catch (...)
             {
-                IException *E = MakeStringException(ROXIE_INTERNAL_ERROR, "Unknown exception caught in CRoxieServerThroughSpillActivity::start");
+                IException *E = MakeStringException(ROXIE_INTERNAL_ERROR, "Unknown exception caught in CRoxieServerThroughSpillActivity::doStart");
```
Comment (afishbeck) [inline roxie/ccd/ccdserver.cpp:9156]: In this case I don't think the message should change.  One of only a few places where you still override start..

File: roxie/ccd/ccdserver.cpp [hunk 2]
```diff
@@ -16424,7 +16442,7 @@ class CRoxieServerLibraryCallActivity : public CRoxieServerActivity
             }
             catch (...)
             {
-                IException *E = MakeStringException(ROXIE_INTERNAL_ERROR, "Unknown exception caught in CRoxieServerLibraryCallActivity::start");
+                IException *E = MakeStringException(ROXIE_INTERNAL_ERROR, "Unknown exception caught in CRoxieServerLibraryCallActivity::doStart");
```
Comment (afishbeck) [inline roxie/ccd/ccdserver.cpp:16445]: Again I don't think the message should change here.

File: roxie/ccd/ccdserver.cpp [hunk 3]
```diff
@@ -1756,11 +1757,24 @@ class CRoxieServerActivity : implements CInterfaceOf<IRoxieServerActivity>, impl
...
+        else
+            doStart(parentExtractSize, parentExtract, paused);
+    }
+
 protected:
     RecordTranslationMode getEnableFieldTranslation() const
     {
         return factory->getEnableFieldTranslation();
     }
+
```
Comment (afishbeck) [inline roxie/ccd/ccdserver.cpp:1777]: Extra line?

**Review by afishbeck — COMMENTED** [review]
A couple of trivial notes about exception messages.  Otherwise looks good.

File: roxie/ccd/ccdserver.cpp [hunk 4]
```diff
@@ -9138,10 +9153,12 @@ class CRoxieServerThroughSpillActivity : public CRoxieServerActivity
             }
             catch (...)
             {
-                IException *E = MakeStringException(ROXIE_INTERNAL_ERROR, "Unknown exception caught in CRoxieServerThroughSpillActivity::start");
+                IException *E = MakeStringException(ROXIE_INTERNAL_ERROR, "Unknown exception caught in CRoxieServerThroughSpillActivity::doStart");
                 startError.set(E);
                 throw E;
             }
+            if (timeActivities)
+                stats.addStatistic(StTimeStart, timer.elapsedCycles());
```
Comment (ghalliday) [inline roxie/ccd/ccdserver.cpp:9161]: Should be StCyclesTimeStartCycle

File: roxie/ccd/ccdserver.cpp [hunk 5]
```diff
@@ -16454,6 +16472,9 @@ class CRoxieServerLibraryCallActivity : public CRoxieServerActivity
                 Owned<IFinalRoxieInput> output = graph->selectOutput(numInputs+extra.unusedOutputs.item(i3));
                 output->stopall();
             }
+
+            if (timeActivities)
+                stats.addStatistic(StTimeStart, timer.elapsedCycles());
```
Comment (ghalliday) [inline roxie/ccd/ccdserver.cpp:16477]: Should be StCyclesTimeStartCycle

**Comment by ghalliday** [comment]
Note: Time for library calls isn't accurate.  I think it is only the time to initialise the inputs, not the activities within the library.  That needs addressing...

File: system/jlib/jstatcodes.h [hunk 6]
```diff
@@ -245,6 +245,8 @@ enum StatisticKind
     StCostCompile,
     StTimeDependencies,
     StCycleDependenciesCycles,
+    StTimeStart,
```
Comment (afishbeck) [inline system/jlib/jstatcodes.h:248]: Does your previous change mean that StTimeStart is no longer being used?

**Review by afishbeck — APPROVED** [review]

**Comment by richardkchapman** [comment]
Little nervous at the scope of the change - one missed parent::start change and you would have infinite recursion I think.

Are you sure this is the least intrusive way of doing it?

**Comment by ghalliday** [comment]
> Little nervous at the scope of the change - one missed parent::start change and you would have infinite recursion I think.
> 
> Are you sure this is the least intrusive way of doing it?

I couldn't think of a better approach - because it needs to cover the whole of the start() function, not just the base class.  When I implemented it I made start private in the base class to make sure I had caught everything.  I could reinstate that, but there are a couple of calls which need some slightly weird coding to get around that change.
- calls from execute()
- call from executeChild()
- calls to CRoxieServerNWayInputBaseActivity->start (can just call cur->) instead.

**Pull request merged**